File: core/ByoSnapCSharp/Controllers/ExampleController.cs
using Microsoft.AspNetCore.Mvc;
using ByoSnapCSharp.Filters;
using ByoSnapCSharp.Models;
using ByoSnapCSharp.Utilities;
using Swashbuckle.AspNetCore.Annotations;

namespace ByoSnapCSharp.Controllers
{
  // =========================================================================
  // Regular API Endpoints — your Snap's business logic lives here.
  //
  // The stubs below demonstrate each Snapser auth exposure. The
  // [SnapserAuth(...)] attribute drives the `x-snapser-auth-types` tag in the
  // generated swagger (which controls which SDK / tool the API surfaces in),
  // and the matching [ValidateAuthorization(...)] enforces it at runtime. Add,
  // rename, or remove these to fit your Snap.
  //
  // A single endpoint can accept MULTIPLE auth types at once (see the last
  // example) — you do not need a separate route per auth type.
  // =========================================================================

  [ApiController]
  [Route("v1/byosnap-core")]
  [Produces("application/json")]
  public class ExampleController : ControllerBase
  {
    private readonly ILogger<ExampleController> _logger;

    public ExampleController(ILogger<ExampleController> logger)
    {
      _logger = logger;
    }

    /// <summary>
    /// Example endpoint exposed over User auth. Accessible by a logged-in user,
    /// validated against the User-Id in their token. Surfaces in the client/game SDK.
    /// </summary>
    [HttpGet("users/{UserId}/example")]
    [SnapserAuth(AppConstants.userAuthType)]
    [ValidateAuthorization(AppConstants.userAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessMessageSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "ExampleUserAuth",
      Summary = "Example: User Auth",
      Description = "Accessible by a logged-in user, validated against the User-Id in their token. Surfaces in the client/game SDK."
    )]
    public ActionResult ExampleUserAuth([FromRoute] UserIdParameterSchema userParams)
    {
      // TODO: Add your user-scoped business logic here.
      //       See advanced/ByoSnapCSharp for the full implementation.
      return Ok(new SuccessMessageSchema
      {
        Message = $"Hello user {userParams.UserId}"
      });
    }

    /// <summary>
    /// Example endpoint exposed over Api-Key auth. Accessible with a valid API
    /// key. Use for trusted server-to-server calls.
    /// </summary>
    [HttpGet("example/api-key")]
    [SnapserAuth(AppConstants.apiKeyAuthType)]
    [ValidateAuthorization(AppConstants.apiKeyAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessMessageSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "ExampleApiKeyAuth",
      Summary = "Example: Api-Key Auth",
      Description = "Accessible with a valid API key. Use for trusted server-to-server calls."
    )]
    public ActionResult ExampleApiKeyAuth()
    {
      // TODO: Add your api-key-scoped business logic here.
      //       See advanced/ByoSnapCSharp for the full implementation.
      return Ok(new SuccessMessageSchema
      {
        Message = "Hello api-key caller"
      });
    }

    /// <summary>
    /// Example endpoint exposed over Internal auth. Callable only by other Snaps
    /// within the same Snapend (internal gateway). Surfaces in the internal SDK.
    /// </summary>
    [HttpGet("example/internal")]
    [SnapserAuth(AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessMessageSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "ExampleInternalAuth",
      Summary = "Example: Internal Auth",
      Description = "Callable only by other Snaps within the same Snapend (internal gateway). Surfaces in the internal SDK."
    )]
    public ActionResult ExampleInternalAuth()
    {
      // TODO: Add your internal-only business logic here.
      //       See advanced/ByoSnapCSharp for the full implementation.
      return Ok(new SuccessMessageSchema
      {
        Message = "Hello internal caller"
      });
    }

    /// <summary>
    /// Example endpoint surfaced in the special Admin SDK.
    ///
    /// Note: `admin` is NOT an auth type. Tagging an endpoint with `admin` only
    /// makes it surface in the Admin SDK (used by admin tooling / the Snapser
    /// dashboard). The request itself still arrives through the internal gateway,
    /// so this is guarded with the INTERNAL check.
    /// </summary>
    [HttpGet("example/admin")]
    [SnapserAuth("admin")]
    [ValidateAuthorization(AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessMessageSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "ExampleAdminSdk",
      Summary = "Example: Admin SDK",
      Description = "Surfaces in the Admin SDK for admin tooling / the Snapser dashboard. `admin` controls SDK exposure, not authentication."
    )]
    public ActionResult ExampleAdminSdk()
    {
      // Admin-SDK calls reach the Snap through the internal gateway, so we guard
      // this with the INTERNAL check. The `admin` tag above is what makes this
      // API surface in the Admin SDK.
      // TODO: Add your admin-only business logic here.
      //       See advanced/ByoSnapCSharp for the full implementation.
      return Ok(new SuccessMessageSchema
      {
        Message = "Hello admin caller"
      });
    }

    /// <summary>
    /// Example endpoint that accepts MULTIPLE auth types on ONE route. One
    /// endpoint reachable by a logged-in user, a valid API key, or an internal
    /// Snap. List every auth type you want to allow — no need for a separate
    /// route per type.
    /// </summary>
    [HttpGet("users/{UserId}/example/multi-auth")]
    [SnapserAuth(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessMessageSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(
      OperationId = "ExampleMultiAuth",
      Summary = "Example: Multi Auth",
      Description = "One endpoint reachable by a logged-in user, a valid API key, or an internal Snap. List every auth type you want to allow — no need for a separate route per type."
    )]
    public ActionResult ExampleMultiAuth([FromRoute] UserIdParameterSchema userParams)
    {
      // Pass every accepted auth type to both the [SnapserAuth(...)] tag (for SDK
      // exposure) and the [ValidateAuthorization(...)] attribute (for runtime
      // enforcement).
      // TODO: Add your business logic here.
      //       See advanced/ByoSnapCSharp for the full implementation.
      return Ok(new SuccessMessageSchema
      {
        Message = $"Hello, request for user {userParams.UserId} passed multi-auth"
      });
    }
  }
}

File: core/byosnap-node-ts/src/controllers/exampleController.ts
import { ErrorResponse, SuccessMessageResponse } from '../types/responses';
import { authMiddleware } from '../middleware/authMiddleware';
import { Request as ExpressRequest } from 'express';
import { Controller, Route, Get, Path, Extension, Middlewares, TsoaResponse, Response, Res, Request } from 'tsoa';

// # @GOTCHAS 👋 - Please read GOTCHAS.md

// ===========================================================================
// Example Controller — your Snap's business logic lives here.
//
// The stubs below demonstrate each Snapser auth exposure. The
// `x-snapser-auth-types` extension controls which SDK / tool the API surfaces
// in, and the matching `authMiddleware([...])` enforces it at runtime. Add,
// rename, or remove these to fit your Snap.
//
// A single endpoint can accept MULTIPLE auth types at once (see the last
// example) — you do not need a separate route per auth type.
//
// For complete, working reference implementations see advanced/byosnap-node-ts.
// ===========================================================================

@Route('/v1/byosnap-core')
export class ExampleController extends Controller {

    /**
     * @summary Example: User Auth
     */
    @Get("users/{userId}/example")
    @Extension("x-description", 'Accessible by a logged-in user, validated against the User-Id in their token. Surfaces in the client/game SDK.')
    @Extension("x-snapser-auth-types", ["user"])
    @Response<SuccessMessageResponse>(200, "Success")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["user"])])
    public async exampleUserAuth(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<SuccessMessageResponse> {
        // TODO: Add your user-scoped business logic here.
        //       See advanced/byosnap-node-ts for a full implementation.
        return { message: `Hello user ${userId}` };
    }

    /**
     * @summary Example: Api-Key Auth
     */
    @Get("example/api-key")
    @Extension("x-description", 'Accessible with a valid API key. Use for trusted server-to-server calls.')
    @Extension("x-snapser-auth-types", ["api-key"])
    @Response<SuccessMessageResponse>(200, "Success")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["api-key"])])
    public async exampleApiKeyAuth(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Request() req: ExpressRequest,
    ): Promise<SuccessMessageResponse> {
        // TODO: Add your api-key-scoped business logic here.
        //       See advanced/byosnap-node-ts for a full implementation.
        return { message: 'Hello api-key caller' };
    }

    /**
     * @summary Example: Internal Auth
     */
    @Get("example/internal")
    @Extension("x-description", 'Callable only by other Snaps within the same Snapend (internal gateway). Surfaces in the internal SDK.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<SuccessMessageResponse>(200, "Success")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async exampleInternalAuth(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Request() req: ExpressRequest,
    ): Promise<SuccessMessageResponse> {
        // TODO: Add your internal-only business logic here.
        //       See advanced/byosnap-node-ts for a full implementation.
        return { message: 'Hello internal caller' };
    }

    /**
     * @summary Example: Admin SDK
     */
    @Get("example/admin")
    @Extension("x-description", 'Surfaces in the Admin SDK for admin tooling / the Snapser dashboard. `admin` controls SDK exposure, not authentication.')
    @Extension("x-snapser-auth-types", ["admin"])
    @Response<SuccessMessageResponse>(200, "Success")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async exampleAdminSdk(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Request() req: ExpressRequest,
    ): Promise<SuccessMessageResponse> {
        // NOTE: `admin` is NOT an auth type. Tagging an endpoint with `admin` in
        // x-snapser-auth-types only makes it surface in the Admin SDK; the request
        // itself still arrives through the internal gateway, so we guard it with
        // the INTERNAL auth check.
        // TODO: Add your admin-only business logic here.
        //       See advanced/byosnap-node-ts for a full implementation.
        return { message: 'Hello admin caller' };
    }

    /**
     * @summary Example: Multi Auth
     */
    @Get("users/{userId}/example/multi-auth")
    @Extension("x-description", 'One endpoint reachable by a logged-in user, a valid API key, or an internal Snap. List every auth type you want to allow - no need for a separate route per type.')
    @Extension("x-snapser-auth-types", ["user", "api-key", "internal"])
    @Response<SuccessMessageResponse>(200, "Success")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["user", "api-key", "internal"])])
    public async exampleMultiAuth(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<SuccessMessageResponse> {
        // One endpoint can accept multiple auth types. Pass every accepted auth
        // type to both the x-snapser-auth-types extension (for SDK exposure) and
        // the authMiddleware (for runtime enforcement).
        // TODO: Add your business logic here.
        //       See advanced/byosnap-node-ts for a full implementation.
        return { message: `Hello, request for user ${userId} passed multi-auth` };
    }
}
